contracts/WitnetProxy.sol

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.5;
+
+import "./utils/Upgradable.sol";
+
+contract WitnetProxy {
+  Upgradable public wrb;
+
+  constructor () {}
+
+  receive() external payable {
+    revert("WitnetProxy: no ETH accepted");
+  }
+
+  // solhint-disable-next-line
+  fallback() external payable {
+    address _wrb = address(wrb);
+    assembly {
+      let ptr := mload(0x40)
+      calldatacopy(ptr, 0, calldatasize())
+      // solhint-disable-next-line
+      let result := delegatecall(gas(), _wrb, ptr, calldatasize(), 0, 0)
+      let size := returndatasize()
+      returndatacopy(ptr, 0, size)
+      switch result
+        case 0  { revert(ptr, size) }
+        default { return(ptr, size) }
+    }
+  }
+
+  function upgrade(address _newWrb, bytes memory _initData) public {
```

Comment:
I can't find the `onlyOwner` check. Does that mean that anyone can upgrade the `WitnetProxy`?

Reply:
So right. `upgrade` function should delegate call to `Upgradable(delegate).canUpgradeFrom(msg.sender)`, instead of surrogate call to `Upgradable(delegate).isUpgradable()`, to assert whether the contract if upgradable from `msg.sender`, before delegating call to `Upgradable(newDelegate).initialize(bytes)`.

Note: new implementations can reuse stored data from previous implementations, as they all run in the context of the proxy address, so it is up to new implementations to require "ownership" or any other role from `msg.sender` when processing the `initialize(bytes)` method.

Reply:
New unitary tests checking these situations will be added.